### PR TITLE
feat: new error types for destinations (snapchat_custom_audience, snapchat_conversion, slack, singular, signl4)

### PR DESCRIPTION
**Pull request by mihir-4116**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/singular/util.js [hunk 1]
```diff
@@ -106,7 +109,7 @@ const platformWisePayloadGenerator = (message, isSessionEvent) => {
   let platform = getValueFromMessage(message, "context.os.name");
   const typeOfEvent = isSessionEvent ? "SESSION" : "EVENT";
   if (!platform) {
-    throw new CustomError("[Singular] :: Platform name is missing", 400);
+    throw new InstrumentationError("Platform name is missing", 400);
```
Comment (koladilip) [inline src/v0/destinations/singular/util.js:112]: error status is not required right

File: src/v0/destinations/slack/transform.js [hunk 2]
```diff
@@ -179,7 +179,7 @@ function processTrack(message, destination) {
   const eventTemplateConfig = destination.Config.eventTemplateSettings;
 
   if (!message.event) {
-    throw new CustomError("Event name is required", 400);
+    throw new InstrumentationError("Event name is required", 400);
```
Comment (koladilip) [inline src/v0/destinations/slack/transform.js:182]: error status is not required right

**Review by koladilip — APPROVED** [review]

**Review by saikumarrs — APPROVED** [review]

File: src/v0/destinations/singular/util.js (hunk 1 above)
Comment (mihir-4116) [reply src/v0/destinations/singular/util.js]: updated

File: src/v0/destinations/slack/transform.js (hunk 2 above)
Comment (mihir-4116) [reply src/v0/destinations/slack/transform.js]: updated

**Review by saikumarrs — APPROVED** [review]

File: src/v0/destinations/singular/util.js [hunk 3]
```diff
@@ -115,7 +118,7 @@ const platformWisePayloadGenerator = (message, isSessionEvent) => {
   }
   platform = platform.toLowerCase();
   if (!SUPPORTED_PLATFORM[platform]) {
-    throw new CustomError("[Singular] :: Platform is not supported");
+    throw new InstrumentationError("Platform is not supported");
```
Comment (utsabc) [inline src/v0/destinations/singular/util.js:121]:
```suggestion
    throw new InstrumentationError(`Platform ${platform} is not supported`);
```

File: src/v0/destinations/singular/util.js [hunk 4]
```diff
@@ -106,7 +109,7 @@ const platformWisePayloadGenerator = (message, isSessionEvent) => {
   let platform = getValueFromMessage(message, "context.os.name");
   const typeOfEvent = isSessionEvent ? "SESSION" : "EVENT";
   if (!platform) {
-    throw new CustomError("[Singular] :: Platform name is missing", 400);
+    throw new InstrumentationError("Platform name is missing");
```
Comment (utsabc) [inline src/v0/destinations/singular/util.js:112]:
```suggestion
    throw new InstrumentationError("Platform name is missing from context.os.name");
```

File: src/v0/destinations/snapchat_conversion/transform.js [hunk 5]
```diff
@@ -57,23 +57,16 @@ function trackResponseBuilder(message, { Config }, mappedEvent) {
     (eventConversionType === "WEB" || eventConversionType === "OFFLINE") &&
     !pixelId
   ) {
-    throw new CustomError(
-      "[Snapchat] :: Pixel Id is required for web and offline events",
-      400
+    throw new InstrumentationError(
```
Comment (utsabc) [inline src/v0/destinations/snapchat_conversion/transform.js:60]: This should be `configuration error`

File: src/v0/destinations/snapchat_conversion/transform.js [hunk 6]
```diff
@@ -57,23 +57,16 @@ function trackResponseBuilder(message, { Config }, mappedEvent) {
...
     );
   }
 
   if (eventConversionType === "MOBILE_APP" && !(appId && snapAppId)) {
     if (!appId) {
-      throw new CustomError(
-        "[Snapchat] :: App Id is required for app events",
-        400
-      );
+      throw new InstrumentationError("App Id is required for app events");
```
Comment (utsabc) [inline src/v0/destinations/snapchat_conversion/transform.js:67]: `configuration error`

File: src/v0/destinations/snapchat_conversion/transform.js [hunk 7]
```diff
@@ -57,23 +57,16 @@ function trackResponseBuilder(message, { Config }, mappedEvent) {
...
     } else {
-      throw new CustomError(
-        "[Snapchat] :: Snap App Id is required for app events",
-        400
-      );
+      throw new InstrumentationError("Snap App Id is required for app events");
```
Comment (utsabc) [inline src/v0/destinations/snapchat_conversion/transform.js:69]: Same here

File: src/v0/destinations/snapchat_custom_audience/transform.js [hunk 8]
```diff
@@ -100,7 +99,7 @@ const getHashedProperty = (schema, hashedProperty) => {
     case "mobileAdId":
       return sha256(hashedProperty.toLowerCase());
     default:
-      throw new CustomError("Invalid schema", 400);
+      throw new InstrumentationError("Invalid schema");
```
Comment (utsabc) [inline src/v0/destinations/snapchat_custom_audience/transform.js:102]: Config error

File: src/v0/destinations/snapchat_custom_audience/transform.js [hunk 9]
```diff
@@ -80,7 +79,7 @@ const getProperty = (schema, element) => {
         element?.MOBILE_ID;
       return property;
     default:
-      throw new CustomError("Invalid schema", 400);
+      throw new InstrumentationError("Invalid schema");
```
Comment (utsabc) [inline src/v0/destinations/snapchat_custom_audience/transform.js:82]: Config error

File: src/v0/destinations/snapchat_custom_audience/utils.js [hunk 10]
```diff
@@ -7,56 +7,36 @@ const ErrorBuilder = require("../../util/error");
...
 const validateFields = (schema, data) => {
   // if required field is not present in all the cases
   if (data[0].length === 0) {
-    throw new ErrorBuilder()
-      .setMessage(
-        `[snapchat_custom_audience]::${schema} is required for the chosen schema.`
-      )
-      .setStatus(400)
-      .build();
+    throw new InstrumentationError(
```
Comment (utsabc) [inline src/v0/destinations/snapchat_custom_audience/utils.js:37]: Configuration error

File: src/v0/destinations/snapchat_custom_audience/utils.js [hunk 11]
```diff
@@ -34,9 +37,7 @@ const validatePayload = message => {
 const validateFields = (schema, data) => {
   // if required field is not present in all the cases
   if (data[0].length === 0) {
-    throw new InstrumentationError(
-      `${schema} is required for the chosen schema`
-    );
+    throw new ConfigurationError(`${schema} is required for the chosen schema`);
```
Comment (saikumarrs) [inline src/v0/destinations/snapchat_custom_audience/utils.js:40]: Somehow this error message doesn't make sense. Can you fix it?

Comment (mihir-4116) [reply src/v0/destinations/snapchat_custom_audience/utils.js]: i made change in sentence, can you confirm it?

**Review by saikumarrs — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Pull request merged**